From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tariq Bashir <tbashir@brightsign.biz>
Date: Thu, 12 Jun 2025 12:30:02 +0100
Subject: Add switch to disable font matching cache: OS-14352

This patch adds a switch to disable font family matching cache in FontServiceApp. This cache
is created and lives in the main process and is not cleared as its a LRU cache. This causes
issues when the AddFont API is called in between using multiple roHtmlWidgets, as the cache
isn't cleared as the main process isn't restarted.

An analysis was made on how to clear the cache when the AddFont API is called from Electron,
but this was found to be difficult and not worth the effort:
The way that FontServiceApp, which runs in the main process, is accessed is via FontLoader,
which is called in the render process. We call AddFont in the main process and so cannot use
FontLoader to call a new CacheClear method in FontServiceApp. Also, trying to get a reference
and directly call FontServiceApp would be very messy. Finally, since we call AddFont before the
render process has been created, to add a call from there triggered by Electron is also not easy.

diff --git a/base/base_switches.cc b/base/base_switches.cc
index f67fc3b6bfe03daa69a5d37a719808f2e05af8cb..19469263a948ea794d135d64084af34a68f10bd4 100644
--- a/base/base_switches.cc
+++ b/base/base_switches.cc
@@ -8,6 +8,11 @@
 
 namespace switches {
 
+// Disables font family matching cache in FontServiceApp. This is required for
+// when using the AddFont API to add a font in FontConfig as the cache lives
+// as long as the main process is running and isn't manually cleared.
+const char kDisableFontFamilyMatchingCache[] = "disable-font-family-matching-cache";
+
 // Delays execution of TaskPriority::BEST_EFFORT tasks until shutdown.
 const char kDisableBestEffortTasks[] = "disable-best-effort-tasks";
 
diff --git a/base/base_switches.h b/base/base_switches.h
index 9764270e00b56f8734fb2079a66b217ea89e90fd..ddaeb0ab2b5ac9a346378b2fe23f25d7f735b341 100644
--- a/base/base_switches.h
+++ b/base/base_switches.h
@@ -14,6 +14,7 @@ namespace switches {
 extern const char kDisableBestEffortTasks[];
 extern const char kDisableBreakpad[];
 extern const char kDisableFeatures[];
+extern const char kDisableFontFamilyMatchingCache[];
 extern const char kDisableLowEndDeviceMode[];
 extern const char kEnableCrashReporter[];
 extern const char kEnableFeatures[];
diff --git a/components/services/font/font_service_app.cc b/components/services/font/font_service_app.cc
index 0c1554c9d3e7375b139f3853247ccb4401a136ca..94be1f10c861543c865dea9885cf317dbce5e7ff 100644
--- a/components/services/font/font_service_app.cc
+++ b/components/services/font/font_service_app.cc
@@ -6,6 +6,7 @@
 
 #include <utility>
 
+#include "base/base_switches.h"
 #include "base/command_line.h"
 #include "base/feature_list.h"
 #include "base/files/file.h"
@@ -146,12 +147,14 @@ void FontServiceApp::MatchFamilyName(const std::string& family_name,
     style->slant = static_cast<mojom::TypefaceSlant>(SkFontStyle().slant());
   }
 
-  // Add to the cache.
-  MatchCacheValue value;
-  value.family_name = result_family_cppstring;
-  value.identity = identity ? identity.Clone() : nullptr;
-  value.style = style.Clone();
-  match_cache_.Put(key, std::move(value));
+  if (!base::CommandLine::ForCurrentProcess()->HasSwitch(switches::kDisableFontFamilyMatchingCache)) {
+    // Add to the cache.
+    MatchCacheValue value;
+    value.family_name = result_family_cppstring;
+    value.identity = identity ? identity.Clone() : nullptr;
+    value.style = style.Clone();
+    match_cache_.Put(key, std::move(value));
+  }
 
   std::move(callback).Run(std::move(identity), result_family_cppstring,
                           std::move(style));
